grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandlerBuilder.java

```diff
@@ -0,0 +1,245 @@
+     * If a custom JSON marshaller has already been set, calling this method will
+     * result in an {@link IllegalStateException}.
+     *
+     * @param messages The collection of messages to register
+     */
+    public UnframedGrpcErrorHandlerBuilder registerMarshalledMessages(
+            Iterable<? extends Message> messages) {
+        requireNonNull(messages, "messages");
+        checkState(jsonMarshaller == null,
+                   "Cannot register the collection of messages because a custom JSON marshaller has " +
+                   "already been set. Use the custom marshaller to register custom messages."
+        );
+        if (marshalledMessages == null) {
+            marshalledMessages = new ArrayList<>();
+        }
+
+        marshalledMessages.addAll(ImmutableList.copyOf(messages));
+
+        return this;
+    }
+
+    /**
+     * Registers custom message types to be marshalled by the error handler.
+     *
+     * <p>This method registers specific message types for custom error responses.
+     * If a custom JSON marshaller has already been set, calling this method will
+     * result in an {@link IllegalStateException}.
+     *
+     * @param messageTypes The message types to register
+     */
+    @SafeVarargs
+    public final UnframedGrpcErrorHandlerBuilder registerMarshalledMessageTypes(
+            Class<? extends Message>... messageTypes) {
+        requireNonNull(messageTypes, "messageTypes");
+        checkState(jsonMarshaller == null,
+                   "Cannot register custom messageTypes because a custom JSON marshaller has already been " +
+                   "set. Use the custom marshaller to register custom message types."
+        );
+
+        if (marshalledMessageTypes == null) {
+            marshalledMessageTypes = new ArrayList<>();
+        }
+        marshalledMessageTypes.addAll(Arrays.asList(messageTypes));
+        return this;
+    }
+
+    /**
+     * Registers custom message types to be marshalled by the error handler.
+     *
+     * <p>This method allows registering message instances for custom error responses.
+     * If a custom JSON marshaller has already been set, calling this method will
+     * result in an {@link IllegalStateException}.
+     *
+     * @param messageTypes The collection of message types to register
+     */
+    public UnframedGrpcErrorHandlerBuilder registerMarshalledMessageTypes(
+            Iterable<? extends Class<? extends Message>> messageTypes) {
+        requireNonNull(messageTypes, "messageTypes");
+        checkState(jsonMarshaller == null,
+                   "Cannot register the collection of messageTypes because a custom JSON marshaller has " +
+                   "already been set. Use the custom marshaller to register custom message types."
+        );
+        if (marshalledMessageTypes == null) {
+            marshalledMessageTypes = new ArrayList<>();
+        }
+        marshalledMessageTypes.addAll(ImmutableList.copyOf(messageTypes));
```

Comment:
Could be simplified into:
```suggestion
        messageTypes.forEach(marshalledMessageTypes::add);
```